File: frontend/src/types/canonicalSet.ts
import type { CanonicalItemBase } from "./canonicalItem";
import type { CanonicalModifierReference } from "./canonicalModifier";
import type { UniqueSetSpecialMechanicClassification } from "./canonicalUnique";

export interface CanonicalSetItem extends CanonicalItemBase {
  set_id?: string | null;
  set_group_id?: string | null;
  set_group_display_name?: string | null;
  base_item_id?: string | null;
  modifier_references: CanonicalModifierReference[];
  modifier_rows?: Array<Record<string, unknown>>;
  tooltip_text?: string[];
  special_mechanic_classification?: UniqueSetSpecialMechanicClassification;
  stable_calculable?: boolean;
}

export interface CanonicalSet extends CanonicalItemBase {
  set_group_id?: string | null;
  set_group_display_name?: string | null;
  item_ids: string[];
  bonus_ids?: string[];
  bonus_modifier_references: CanonicalModifierReference[];
  bonus_text: string[];
  special_mechanic_classification?: UniqueSetSpecialMechanicClassification;
  stable_calculable?: boolean;
}

export interface CanonicalSetBonus {
  canonical_id: string;
  display_name: string;
  set_group_id: string;
  set_group_display_name?: string | null;
  required_pieces?: number | null;
  support_status: string;
  trust_level: string;
  description_text: string[];
  modifier_references: CanonicalModifierReference[];
  modifier_rows?: Array<Record<string, unknown>>;
  special_mechanic_classification: UniqueSetSpecialMechanicClassification;
  stable_calculable?: boolean;
}

File: backend/app/data_contracts/canonical_set.py
"""Canonical set item contracts."""

from __future__ import annotations

from dataclasses import dataclass, field

from .canonical_item import CanonicalItemBase
from .canonical_modifier import CanonicalModifierReference


@dataclass(frozen=True)
class CanonicalSetItem(CanonicalItemBase):
    set_id: str | None = None
    base_item_id: str | None = None
    modifier_references: tuple[CanonicalModifierReference, ...] = field(default_factory=tuple)

    def __post_init__(self) -> None:
        super().__post_init__()
        object.__setattr__(self, "modifier_references", tuple(self.modifier_references))


@dataclass(frozen=True)
class CanonicalSet(CanonicalItemBase):
    item_ids: tuple[str, ...] = field(default_factory=tuple)
    bonus_modifier_references: tuple[CanonicalModifierReference, ...] = field(default_factory=tuple)
    bonus_text: tuple[str, ...] = field(default_factory=tuple)

    def __post_init__(self) -> None:
        super().__post_init__()
        object.__setattr__(self, "item_ids", tuple(self.item_ids))
        object.__setattr__(self, "bonus_modifier_references", tuple(self.bonus_modifier_references))
        object.__setattr__(self, "bonus_text", tuple(self.bonus_text))
